peps/pep-0733.rst

```diff
@@ -0,0 +1,629 @@
+  having to create deep copies.
+* Provide and import frozen modules.
+* Run multiple independent interpreters (in particular, when embedded
+  in a library that wants to avoid global effects).
+
+Alternative Python Implementations
+----------------------------------
+
+Alternative implementations of Python (such as
+`PyPy <https://www.pypy.org>`__,
+`GraalPy <https://www.graalvm.org/python/>`__,
+`IronPython <https://ironpython.net>`__,
+`RustPython <https://github.com/RustPython/RustPython>`__,
+`MicroPython <https://micropython.org>`__,
+and `Jython <https://www.jython.org>`__), may take
+very different approaches for the implementation of
+different subsystems. They need:
+
+* The API to be abstract and hide implementation details.
+* A specification of the API, ideally with a test suite
+  that ensures compatibility.
+* It would be nice to have an ABI that can be shared
+  across Python implementations.
+
+Alternative APIs
+----------------
+
+There are several projects that implement alternatives to the
+C API, which offer extension users advantanges over programming
+directly with the C API. These APIs are implemented with the
+C API, and in some cases by using CPython internals.
+Some examples are
+`Cython <https://cython.org>`__,
+`HPy <https://hpyproject.org>`__ and
+`pythoncapi-compat <https://pythoncapi-compat.readthedocs.io/en/latest/>`__.
+CPython's DSL for parsing function arguments, the
+`Argument Clinic <https://docs.python.org/3/howto/clinic.html>`__,
+can also be seen as belonging to this category of stakeholders.
+
+Such systems need minimal building blocks for accessing CPython
+efficiently. They don't necessarily need an ergonomic API, because
+they typically generate code that is not intended to be read
+by humans. But they do need it to be comprehensive enough so that
+they don't need to access internals, while offering them stability,
+and without sacrificing performance.
+
+An alternative is to have a fast API tier with less error checking
+and lower stability guarantees. Then the developers and users of
+these tools can choose whether to generate code that uses the
+faster or the safer and more stable version of the API.
+
+Binding Generators
+------------------
+
+Libraries that create bindings between Python and other object models,
+paradigms or languages, such as
+`pybind11 <https://pybind11.readthedocs.io/en/stable/>`__ for C++11,
+`PyO3 <https://github.com/PyO3/pyo3>`__ for Rust,
+`PySide <https://pypi.org/project/PySide/>`__ for Qt,
+`PyGObject <https://pygobject.readthedocs.io/en/latest/>`__ for GTK,
+`Pygolo <https://gitlab.com/pygolo/py>`__ for Go,
+`PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java,
+`SWIG <https://swig.org/>`__ for C/C++, or
+`Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).
+
+Some of the alternative APIs mentioned in the previous section create
+alternative object models and therefore they can also be seen as
+belonging to this category.
+
+They need to:
```

Comment:
Having now read this section, I think `Cython` belongs here and not in the previous one.

Higher-level abstractions over the C API (the previous section) are probably a less important consideration, considering they're largely designed to cover over the flaws/limitations of the C API that we want to fix. We should be aiming to make them mostly redundant, rather than trying to actively support them (e.g. if we had designed it perfectly in the first place, HPy and pythoncapi-compat would never have existed).

Cython is different enough from our C API that we aren't going to make it redundant, so I think it belongs here.

Reply:
IMO, Cython belongs in both. Many of these do, actually.
Since these are non-exhaustive lists, it's OK to only put it in one place.

Reply:
> Since these are non-exhaustive lists, it's OK to only put it in one place.

We're not concerned about the exhaustiveness of the lists, but that Cython is currently in the _wrong_ list. Non-exhaustive doesn't mean you can incorrectly categorise the items that you do include.